include/psa/crypto_extra.h

```diff
@@ -711,6 +711,99 @@ psa_status_t mbedtls_psa_external_get_random(
 
 /**@}*/
 
+/** \defgroup psa_builtin_keys Built-in keys
+ * @{
+ */
+
+/** The minimum value for a key identifier that is built into the
+ * implementation.
+ *
+ * The range of key identifiers from #MBEDTLS_PSA_KEY_ID_BUILTIN_MIN
+ * to #MBEDTLS_PSA_KEY_ID_BUILTIN_MAX within the range from
+ * #PSA_KEY_ID_VENDOR_MIN and #PSA_KEY_ID_VENDOR_MAX and must not intersect
+ * with any other set of implementation-chosen key identifiers.
+ *
+ * This value is part of the library's ABI since changing it would invalidate
+ * the values of built-in key identifiers in applications.
+ */
+#define MBEDTLS_PSA_KEY_ID_BUILTIN_MIN          ((psa_key_id_t)0x7fff0000)
+
+/** The maximum value for a key identifier that is built into the
+ * implementation.
+ *
+ * See #MBEDTLS_PSA_KEY_ID_BUILTIN_MIN for more information.
+ */
+#define MBEDTLS_PSA_KEY_ID_BUILTIN_MAX          ((psa_key_id_t)0x7fffefff)
+
+/** A slot number identifying a key in a driver.
+ *
+ * Values of this type are used to identify built-in keys.
+ */
+typedef uint64_t psa_drv_slot_number_t;
+
+#if defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
+/** Test whether a key identifier belongs to the builtin key range.
+ *
+ * \param key_id  Key identifier to test.
+ *
+ * \retval 1
+ *         The key identifier is a builtin key identifier.
+ * \retval 0
+ *         The key identifier is not a builtin key identifier.
+ */
+static inline int psa_key_id_is_builtin( psa_key_id_t key_id )
+{
+    return( ( key_id >= MBEDTLS_PSA_KEY_ID_BUILTIN_MIN ) &&
+            ( key_id <= MBEDTLS_PSA_KEY_ID_BUILTIN_MAX ) );
+}
+
+/** Platform function to obtain the location and slot of a built-in key.
+ *
+ * An application-specific implementation of this function must be provided if
+ * #MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS is enabled. This would typically be provided
+ * as part of a platform's system image.
+ *
+ * #MBEDTLS_SVC_KEY_ID_GET_KEY_ID(\p key_id) needs to be in the range from
+ * #MBEDTLS_PSA_KEY_ID_BUILTIN_MIN to #MBEDTLS_PSA_KEY_ID_BUILTIN_MAX.
+ *
+ * In a multi-application configuration
+ * (\c MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER is defined),
+ * this function should check that #MBEDTLS_SVC_KEY_ID_GET_OWNER_ID(\p key_id)
+ * is allowed to use the given key.
+ *
+ * \param key_id                The key ID for which to retrieve the
+ *                              location and slot attributes.
+ * \param[out] lifetime         On success, the lifetime associated with the key
```

Comment:
I disagree. The platform should be able to indicate to the core whether a platform-installed key is read-only or modifiable.

Reply:
To be sure we are on the same page here, as I mentioned in my comment, in `psa-driver-interface.md` there is a specific point in the `acme_get_builtin_key` entry point specification that the driver may change the persistence level of the key from PSA_KEY_LIFETIME_PERSISTENT to something else:

```                                                                             
psa_status_t acme_get_builtin_key(psa_drv_slot_number_t slot_number,            
                                  psa_key_attributes_t *attributes,             
                                  uint8_t *key_buffer,                          
                                  size_t key_buffer_size,                       
                                  size_t *key_buffer_length);                   
```                                                                             
                                                                                
If this function returns `PSA_SUCCESS` or `PSA_ERROR_BUFFER_TOO_SMALL`, it must fill `attributes` with the attributes of the key (except for the key identifier). On success, this function must also fill `key_buffer` with the key context.
                                                                                
On entry, `psa_get_key_lifetime(attributes)` is the location at which the driver was declared and the persistence level `#PSA_KEY_LIFETIME_PERSISTENT`. The driver entry point may change the lifetime to one with the same location but a different persistence level. The standard attributes other than the key identifier and lifetime have the value conveyed by `PSA_KEY_ATTRIBUTES_INIT`.

Thus there is already a way for the core to be told if a key is read-only, through the driver that provides the description of the key. Is it the case that you want ALSO the platform function to be able to provide this information, for cases where the driver would not know the persistent level of the key (but all its other attributes ... ).

If we go for that, the specification of `acme_get_builtin_key` needs to be updated as the passed persistent level would not necessarily be `#PSA_KEY_LIFETIME_PERSISTENT`.

Reply:
If I recall correctly a discussion I had with @gilles-peskine-arm about this, there was a wish to support a use-case where a key in hardware might be erasable, but for the platform function (which maps the PSA IDs to a driver location + slot number) to be able to 'override' this, and not allow erase operations to be processed through the PSA Core. That'd mean the platform function would in the end be responsible for setting the PSA-facing lifetime value, and the driver-specific `get_builtin_key` hook would be allowed to 'override' the lifetime in case the platform tries to set it to something the driver doesn't support.

So IMO, the current flow adheres to this train of thought. I'll defer to @gilles-peskine-arm for more comments, and whether or not a doc update is needed.

Reply:
Persistence levels are meant to help key management on devices that have a complex lifecycle. Typically it would indicate which keys must be erased on a factory reset, on a transfer of ownership, on a transfer of tenancy, or other similar events. We lump everything that isn't an application, our own generic library code or code that's closely associated with hardware (drivers) as “the platform”. The platform is where lifecycle-related decisions are made (at least after some initial manufacturing stages). So the platform gets to decide what the persistence level of a key is.

Does this mean that the platform should be able to override the persistence level reported by drivers? Arguably, no: the platform could decide the persistence level, and have the driver store this information and retrieve it. However, a built-in key may be stored in, say, a secure element that doesn't have room for this extra piece of metadata. In such a case, it would be onerous to require that the driver implement some additional metadata storage just to remember the persistence level (the driver could know the key type because that's an intrinsic property of the key, and it could even provide a sensible usage policy which the platform would only need to override if it's defending against a hostile application). So it makes sense to me that the platform could override the persistence level reported by the driver: the driver can just report the default level (or `READ_ONLY` for keys that are physically read-only), and the platform can override this for keys that must be preserved across device lifecycle stages.

I haven't looked at the code of this pull request in detail so I don't know exactly what this means in terms of implementation.

Reply:
The current state of this PR is that the full lifetime value for a key (location + persistence) is declared by the platform hook through mapping the builtin key ID. The lifetime value and key ID are then set onto a fresh key attributes structure, which is then passed onto the driver for completion, just as defined in `psa-driver-interface.md`.

The divergence here is that the current PR state opens up for the platform to declare a key with a different persistence than `PSA_KEY_LIFETIME_PERSISTENT`, which will be set to the key attributes struct before passing it onto `acme_get_builtin_key()`. This matches with the behaviour that you described in an inverse way - i.e. the platform attempts to override the persistency level on a key, and the driver entry point can complain if the platform-declared lifetime is fundamentally incompatible with the actual capabilities of the driver.

Is that acceptable, or does it actually need to happen in the order of driver reports default lifetime of key as known to the driver, and then the platform can override it to a platform-set value? With no validation from the driver-side to see whether the platform-set value is actually possible for the actual key?

I believe the current state allows for the lifetime override by the platform to happen in a sensible way. Do you agree, @gilles-peskine-arm ? In that case, I'll make one more commit to change the markdown documentation and make it clear that the key lifetime may be set to any value on entry to `get_builtin_key`, and the driver should report an error if the registered lifetime is fundamentally impossible for the requested key.

Reply:
If there is no room to store the persistence level together with the key material, the two approaches are mostly equivalent: the only source for the persistent level is inside the platform (whether it's hard-coded or stored in some implementation-specific way). There's even a slight advantage to allowing the driver to validate the persistence level: it can prevent a platform from claiming that a key is erasable when the hardware knows that it isn't. I don't consider this a significant advantage because either way this is the sort of thing that would only be discovered by integration testing, and it's about as easy to test either way.

However, if there is room to store the persistence level with the key material, there is a definite advantage in allowing the driver to let the platform know what the persistence level is. This way, the persistence level can be decided at key provisioning time and stored together with the key's other metadata (type, policy, etc.).

How about keeping the code as it is, but changing the contract to allow the driver to override the lifetime in attributes? So if there's no room in the hardware, the driver just validates the persistence level (which I think only means to insist on `READ_ONLY` if the key is physically read-only). But if the hardware does store the persistence level, the driver may override the value in the attributes, and knowing this the platform can just put a default value and let the driver do its thing.

Reply:
> How about keeping the code as it is, but changing the contract to allow the driver to override the lifetime in attributes?

That's exactly what's implemented and what I proposed to do in my last comment. It sounds like you agree with that approach, so I'll update the doc to match the implementation ASAP.

Reply:
I've been through this. @gilles-peskine-arm thanks for the clarifications about what we mean by platform and its role in key lifetime management. I am happy with the proposed solution.

Reply:
I pushed the doc update in my last commit :)